File: sdks/python/hatchet_sdk/features/filters.py
import asyncio

from hatchet_sdk.clients.rest.api.filter_api import FilterApi
from hatchet_sdk.clients.rest.api_client import ApiClient
from hatchet_sdk.clients.rest.models.v1_create_filter_request import (
    V1CreateFilterRequest,
)
from hatchet_sdk.clients.rest.models.v1_filter import V1Filter
from hatchet_sdk.clients.rest.models.v1_filter_list import V1FilterList
from hatchet_sdk.clients.rest.models.v1_update_filter_request import (
    V1UpdateFilterRequest,
)
from hatchet_sdk.clients.v1.api_client import BaseRestClient, retry
from hatchet_sdk.utils.typing import JSONSerializableMapping


class FiltersClient(BaseRestClient):
    """
    The filters client is a client for interacting with Hatchet's filters API.
    """

    def _fa(self, client: ApiClient) -> FilterApi:
        return FilterApi(client)

    async def aio_list(
        self,
        limit: int | None = None,
        offset: int | None = None,
        workflow_ids: list[str] | None = None,
        scopes: list[str] | None = None,
    ) -> V1FilterList:
        """
        List filters for a given tenant.

        :param limit: The maximum number of filters to return.
        :param offset: The number of filters to skip before starting to collect the result set.
        :param workflow_ids: A list of workflow IDs to filter by.
        :param scopes: A list of scopes to filter by.

        :return: A list of filters matching the specified criteria.
        """
        return await asyncio.to_thread(self.list, limit, offset, workflow_ids, scopes)

    @retry
    def list(
        self,
        limit: int | None = None,
        offset: int | None = None,
        workflow_ids: list[str] | None = None,
        scopes: list[str] | None = None,
    ) -> V1FilterList:
        """
        List filters for a given tenant.

        :param limit: The maximum number of filters to return.
        :param offset: The number of filters to skip before starting to collect the result set.
        :param workflow_ids: A list of workflow IDs to filter by.
        :param scopes: A list of scopes to filter by.

        :return: A list of filters matching the specified criteria.
        """
        with self.client() as client:
            return self._fa(client).v1_filter_list(
                tenant=self.tenant_id,
                limit=limit,
                offset=offset,
                workflow_ids=workflow_ids,
                scopes=scopes,
            )

    @retry
    def get(
        self,
        filter_id: str,
    ) -> V1Filter:
        """
        Get a filter by its ID.

        :param filter_id: The ID of the filter to retrieve.

        :return: The filter with the specified ID.
        """
        with self.client() as client:
            return self._fa(client).v1_filter_get(
                tenant=self.tenant_id,
                v1_filter=filter_id,
            )

    async def aio_get(
        self,
        filter_id: str,
    ) -> V1Filter:
        """
        Get a filter by its ID.

        :param filter_id: The ID of the filter to retrieve.

        :return: The filter with the specified ID.
        """
        return await asyncio.to_thread(self.get, filter_id)

    def create(
        self,
        workflow_id: str,
        expression: str,
        scope: str,
        payload: JSONSerializableMapping | None = None,
    ) -> V1Filter:
        """
        Create a new filter.

        :param workflow_id: The ID of the workflow to associate with the filter.
        :param expression: The expression to evaluate for the filter.
        :param scope: The scope for the filter.
        :param payload: The payload to send with the filter.

        :return: The created filter.
        """
        with self.client() as client:
            return self._fa(client).v1_filter_create(
                tenant=self.tenant_id,
                v1_create_filter_request=V1CreateFilterRequest(
                    workflowId=workflow_id,
                    expression=expression,
                    scope=scope,
                    payload=payload,
                ),
            )

    async def aio_create(
        self,
        workflow_id: str,
        expression: str,
        scope: str,
        payload: JSONSerializableMapping | None = None,
    ) -> V1Filter:
        """
        Create a new filter.

        :param workflow_id: The ID of the workflow to associate with the filter.
        :param expression: The expression to evaluate for the filter.
        :param scope: The scope for the filter.
        :param payload: The payload to send with the filter.

        :return: The created filter.
        """
        return await asyncio.to_thread(
            self.create, workflow_id, expression, scope, payload
        )

    def delete(
        self,
        filter_id: str,
    ) -> V1Filter:
        """
        Delete a filter by its ID.

        :param filter_id: The ID of the filter to delete.
        :return: The deleted filter.
        """
        with self.client() as client:
            return self._fa(client).v1_filter_delete(
                tenant=self.tenant_id,
                v1_filter=filter_id,
            )

    async def aio_delete(
        self,
        filter_id: str,
    ) -> V1Filter:
        """
        Delete a filter by its ID.

        :param filter_id: The ID of the filter to delete.
        :return: The deleted filter.
        """
        return await asyncio.to_thread(self.delete, filter_id)

    def update(
        self,
        filter_id: str,
        updates: V1UpdateFilterRequest,
    ) -> V1Filter:
        """
        Update a filter by its ID.

        :param filter_id: The ID of the filter to delete.
        :param updates: The updates to apply to the filter.
        :return: The updated filter.
        """
        with self.client() as client:
            return self._fa(client).v1_filter_update(
                tenant=self.tenant_id,
                v1_filter=filter_id,
                v1_update_filter_request=updates,
            )

    async def aio_update(
        self,
        filter_id: str,
        updates: V1UpdateFilterRequest,
    ) -> V1Filter:
        """
        Update a filter by its ID.

        :param filter_id: The ID of the filter to delete.
        :param updates: The updates to apply to the filter.
        :return: The updated filter.
        """
        return await asyncio.to_thread(self.update, filter_id, updates)

File: sdks/ruby/src/lib/hatchet/features/filters.rb
# frozen_string_literal: true

module Hatchet
  module Features
    # Filters client for interacting with Hatchet's filters API
    #
    # This class provides a high-level interface for creating, retrieving,
    # listing, updating, and deleting filters in the Hatchet system.
    #
    # @example Listing filters
    #   filters = filters_client.list(limit: 10, workflow_ids: ["wf-1"])
    #
    # @example Creating a filter
    #   filter = filters_client.create(
    #     workflow_id: "wf-1",
    #     expression: 'input.priority > 5',
    #     scope: "high-priority"
    #   )
    #
    # @since 0.1.0
    class Filters
      # Initializes a new Filters client instance
      #
      # @param rest_client [Object] The configured REST client for API communication
      # @param config [Hatchet::Config] The Hatchet configuration containing tenant_id and other settings
      # @return [void]
      # @since 0.1.0
      def initialize(rest_client, config)
        @rest_client = rest_client
        @config = config
        @filter_api = HatchetSdkRest::FilterApi.new(rest_client)
      end

      # List filters for the current tenant
      #
      # @param limit [Integer, nil] The maximum number of filters to return
      # @param offset [Integer, nil] The number of filters to skip
      # @param workflow_ids [Array<String>, nil] A list of workflow IDs to filter by
      # @param scopes [Array<String>, nil] A list of scopes to filter by
      # @return [Object] A list of filters matching the specified criteria
      # @raise [HatchetSdkRest::ApiError] If the API request fails
      # @example
      #   filters = filters_client.list(limit: 10, workflow_ids: ["wf-1"])
      def list(limit: nil, offset: nil, workflow_ids: nil, scopes: nil)
        @filter_api.v1_filter_list(
          @config.tenant_id,
          {
            limit: limit,
            offset: offset,
            workflow_ids: workflow_ids,
            scopes: scopes
          }
        )
      end

      # Get a filter by its ID
      #
      # @param filter_id [String] The ID of the filter to retrieve
      # @return [Object] The filter details
      # @raise [HatchetSdkRest::ApiError] If the API request fails
      # @example
      #   filter = filters_client.get("filter-123")
      def get(filter_id)
        @filter_api.v1_filter_get(@config.tenant_id, filter_id)
      end

      # Create a new filter
      #
      # @param workflow_id [String] The ID of the workflow to associate with the filter
      # @param expression [String] The CEL expression to evaluate for the filter
      # @param scope [String] The scope for the filter
      # @param payload [Hash, nil] The payload to send with the filter
      # @return [Object] The created filter
      # @raise [HatchetSdkRest::ApiError] If the API request fails
      # @example
      #   filter = filters_client.create(
      #     workflow_id: "wf-1",
      #     expression: 'input.value > 10',
      #     scope: "my-scope",
      #     payload: { threshold: 10 }
      #   )
      def create(workflow_id:, expression:, scope:, payload: nil)
        request = HatchetSdkRest::V1CreateFilterRequest.new(
          workflow_id: workflow_id,
          expression: expression,
          scope: scope,
          payload: payload
        )
        @filter_api.v1_filter_create(@config.tenant_id, request)
      end

      # Delete a filter by its ID
      #
      # @param filter_id [String] The ID of the filter to delete
      # @return [Object] The deleted filter
      # @raise [HatchetSdkRest::ApiError] If the API request fails
      # @example
      #   filters_client.delete("filter-123")
      def delete(filter_id)
        @filter_api.v1_filter_delete(@config.tenant_id, filter_id)
      end

      # Update a filter by its ID
      #
      # @param filter_id [String] The ID of the filter to update
      # @param updates [Hash] The updates to apply to the filter
      # @return [Object] The updated filter
      # @raise [HatchetSdkRest::ApiError] If the API request fails
      # @example
      #   filters_client.update("filter-123", { expression: 'input.value > 20' })
      def update(filter_id, updates)
        update_request = HatchetSdkRest::V1UpdateFilterRequest.new(updates)
        @filter_api.v1_filter_update(@config.tenant_id, filter_id, update_request)
      end
    end
  end
end
